ang/taskRunner.js

```diff
@@ -66,43 +65,49 @@
         return;
       }
 
-      var data = {}
-      data['task_id'] = taskId;
+      const data = {
+        task_id: taskId,
+        async: true,
+      };
 
       if ($scope.isInputValueRequired()) {
         data['input_val'] = $scope.inputValue;
       }
 
-      CRM.alert("Task execution has started", "Task execution", 'info');
-      $scope.isTaskRunning = true;
-
-      CRM.api3("Sqltask", "execute", data).done(function(result) {
-        if (result.values && !result.is_error) {
-          if (result.values.log !== undefined && Array.isArray(result.values.log)) {
-            $scope.resultLogs = result.values.log;
-            $scope.isTaskReturnsEmptyLogs = $scope.resultLogs.length  === 0;
-          } else {
-            $scope.isTaskReturnsEmptyLogs = true;
-          }
-          $scope.isTaskRunning = false;
-          $scope.isShowLogs = true;
-          CRM.alert("Task execution completed", "Task execution", 'success');
-        } else {
-          CRM.alert(result.error_message, ts("Error task execution"), "error");
-          $scope.resultLogs = [ts('Task returns error: ') + result.error_message];
-          $scope.isTaskRunning = false;
-          $scope.isShowLogs = true;
-        }
-        $scope.runButtonText = $scope.ts('Run again');
-        $scope.$apply();
-      }).fail(function() {
-        $scope.runButtonText = $scope.ts('Run again');
-        $scope.resultLogs = ["An unknown error occurred during task execution. Please check your server logs for details before proceeding."];
-        $scope.isTaskRunning = false;
-        $scope.isShowLogs = true;
+      $scope.isShowLogs = true;
+      $scope.resultLogs = [];
+
+      new Promise((resolve, reject) => {
+        CRM.api3("Sqltask", "execute", data).done(resolve).fail(reject);
+      }).then(async (result) => {
+        if (result.is_error !== 0) throw new Error(result.error_message);
+
+        CRM.alert("Task execution added to background queue", "Task queued", "info");
+
+        $scope.isTaskRunning = true;
         $scope.$apply();
-        CRM.alert("Task failed to execute", "Task execution", 'error');
-      });
+
+        while ($scope.isTaskRunning) {
+          await new Promise(resolve => setTimeout(resolve, 2000));
+
+          const [execution] = await CRM.api4("SqltasksExecution", "get", {
```

Comment:
This approach might be a bit too imprecise. When coworker is down or too slow to pick up new QueueItems by the time this fires, we may either get the results of a previous task execution, or no results at all if the task has never been executed before. With bad timing, you could probably also get a task execution triggered by another user or process.

I'm wondering if a better approach might be something like this:
1. Alter `civicrm_sqltasks_execution` to add a `status` OptionGroup field (and make `start_date` NULLable).
2. Add an optional `sqltasks_execution_id` parameter to the `Sqltask.execute`. When the parameter is set, use that row to track logs/errors/etc. rather than creating a new one
3. When executing a task with `async: true`, create a `SqltasksExecution` entity with status `queued` and include the `sqltasks_execution_id` in the `QueueItem`. Return `sqltasks_execution_id` so clients can use it to poll for task results. 
4. When the task execution begins, set the status to `running` and set `start_date`.
5. When the task completes or fails, set the status to `completed` or `failed`.